### PR TITLE
[CARBONDATA-3900] [CARBONDATA-3882] [CARBONDATA-3881] Fix multiple concurrent issues in table status lock and segment lock for SI and maintable

### PR DESCRIPTION
**Pull request by ajantha-bhat**
### Why is this PR needed?
**[CARBONDATA-3900]** Fix maintable load failure in concurrent load and compaction scenario
Main table load flow segment lock is released before updating the table status success.
So, concurrent operation was considering this segment as stale segment (as segment lock is not present) and cleaning it. Leading to unable to get file status exception. 

**[CARBONDATA-3882]** Fix wrong lock and missing Table status lock in some SI flows
 1) In updateLoadMetadataWithMergeStatus, we want update SI table status, but lock is acquired on main table
2) triggerCompaction, updateTableStatusForIndexTables -> table status write is happening without lock
 
**[CARBONDATA-3881]**  Fix concurrent main table compaction and SI load issue
Consider a scenario, where segmentX has loaded to main table but failed to load to SI table. So, while loading another segmentY, we reload failed SI segmentX. this time if the segmentX is compacted in main table and clean files executed on it. SI load will fail and segment id will not be found in segmentMap of SI and it throws exception.

 ### What changes were proposed in this PR?
**for [CARBONDATA-3900]**
release segment lock after main table is updated.

**for [CARBONDATA-3882]**
1) In updateLoadMetadataWithMergeStatus, take a lock on SI table
2) triggerCompaction, updateTableStatusForIndexTables -> add a lock for table status write

**for [CARBONDATA-3881]**
just before reloading the failed SI segment. check if it is valid segment in main table.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No [issue comes in concurrent scenarios]

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3227/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1499/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1517/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3245/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3274/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1538/

**Comment by ajantha-bhat** [comment]
@kunal642 , @akashrn5 @QiangCai : please check it

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1654/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3395/

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 1]
```diff
@@ -440,14 +448,22 @@ object SecondaryIndexUtil {
         val loadFolderDetailsArray = SegmentStatusManager.readLoadMetadata(indexTable
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:448]: reading should be inside of locking

Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: Agree. Modified.

Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: Also handled similar issue in SecondaryIndexUtil#triggerCompaction

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3429/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1687/

**Comment by QiangCai** [comment]
LGTM

**Pull request closed without merging**